### PR TITLE
#4539 - Add Duplicate SIN Validation

### PR DESCRIPTION
**Pull request by tiago-graf**
### Summary

Adds a duplicate validation when manually adding a SIN number to a student in the Ministry Portal.

### Frontend
Updates the AddNewSIN modal component to include new banner and confirmation checkbox.
Updates the SINManagement view to move the save SIN method to the AddNewSIN component (to allow resubmission without closing the modal)

### Backend
Updates the create SIN model to include optional confirmDuplicate property.
Updates the SIN validation service to include new duplicate validation.
Updates the student controller to handle the custom error (duplicate not confirmed).

### E2E
Adds new tests for the different validation scenarios.

image
image

### CODE REVIEW

File: sources/packages/backend/apps/api/src/route-controllers/student/student.aest.controller.ts [hunk 1]
```diff
@@ -347,6 +352,19 @@ export class StudentAESTController extends BaseController {
     if (!studentExists) {
       throw new NotFoundException("Student does not exists.");
     }
+    // Check for duplicate SIN.
```
Comment (weskubo-cgi) [inline sources/packages/backend/apps/api/src/route-controllers/student/student.aest.controller.ts]: **question:** Do we even need to run this validation if `confirmDuplicateSIN` is true?

File: sources/packages/backend/apps/api/src/services/sin-validation/sin-validation.service.ts [hunk 2]
```diff
@@ -55,6 +55,27 @@ export class SINValidationService extends RecordDataModelService<SINValidation>
...
+   * Checks if the provided SIN exists for another student.
+   * Only checks against the current SIN validation record for each student
+   * (the one referenced by student.sinValidation), not historical records.
+   * @param studentId current student id to exclude from the check.
+   * @param sin SIN to check for duplicates.
+   * @returns true if the SIN exists for another student, false otherwise.
+   */
+  async checkDuplicateSIN(studentId: number, sin: string): Promise<boolean> {
+    const normalizedSIN = removeWhiteSpaces(sin);
+    const existingStudent = await this.dataSource
```
Comment (weskubo-cgi) [inline sources/packages/backend/apps/api/src/services/sin-validation/sin-validation.service.ts]: **suggestion:** Is there a way to write this using `.exists()`? Semantically, that's what we're doing here.

File: sources/packages/backend/apps/api/src/route-controllers/student/_tests_/e2e/student.aest.controller.createStudentSINValidation.e2e-spec.ts [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+  let app: INestApplication;
+  let db: E2EDataSources;
+
+  beforeAll(async () => {
+    const { nestApplication, dataSource } = await createTestingAppModule();
+    app = nestApplication;
+    db = createE2EDataSources(dataSource);
+  });
+
+  it("Should reject creation of duplicate SIN when confirmDuplicateSIN is not provided.", async () => {
```
Comment (weskubo-cgi) [inline sources/packages/backend/apps/api/src/route-controllers/student/_tests_/e2e/student.aest.controller.createStudentSINValidation.e2e-spec.ts]: **suggestion:** Add a test case for adding a duplicate SIN for the same student that should pass validation.

File: sources/packages/backend/apps/api/src/route-controllers/student/_tests_/e2e/student.aest.controller.createStudentSINValidation.e2e-spec.ts [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+    };
+
+    // Act/Assert
+    await request(app.getHttpServer())
+      .post(endpoint)
+      .send(payload)
+      .auth(aestUserToken, BEARER_AUTH_TYPE)
+      .expect(HttpStatus.UNPROCESSABLE_ENTITY)
+      .expect((response) => {
+        expect(response.body.message).toContain(
```
Comment (weskubo-cgi) [inline sources/packages/backend/apps/api/src/route-controllers/student/_tests_/e2e/student.aest.controller.createStudentSINValidation.e2e-spec.ts]: **question:** Is there any reason this isn't an exact match?

File: sources/packages/backend/apps/api/src/route-controllers/student/_tests_/e2e/student.aest.controller.createStudentSINValidation.e2e-spec.ts [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+    // Verify the SIN validation was created.
+    const createdSINValidation = await db.sinValidation.findOne({
+      where: { id: createdId },
+      relations: { student: true },
+    });
+    expect(createdSINValidation).toBeDefined();
+    expect(createdSINValidation.student.id).toBe(student.id);
+    expect(createdSINValidation.sin).toBe(uniqueSIN);
+  });
+
```
Comment (dheepak-aot) [inline sources/packages/backend/apps/api/src/route-controllers/student/_tests_/e2e/student.aest.controller.createStudentSINValidation.e2e-spec.ts]: Great level of E2E tests. Thanks for the efforts.

File: sources/packages/backend/apps/api/src/route-controllers/student/student.aest.controller.ts (hunk 1 above)
Comment (tiago-graf) [reply sources/packages/backend/apps/api/src/route-controllers/student/student.aest.controller.ts]: Good point i don't think it needs to run the check since it will always pass the validation

File: sources/packages/web/src/components/common/sin/AddNewSIN.vue [hunk 6]
```diff
@@ -70,30 +88,67 @@ export default defineComponent({
...
+        snackBar.error("Unexpected error while creating a new SIN record.");
+        return false;
+      }
+
+      return true;
     };
 
     const cancel = () => {
       addNewSINForm.value.reset();
+      showDuplicateWarning.value = false;
```
Comment (dheepak-aot) [inline sources/packages/web/src/components/common/sin/AddNewSIN.vue]: Is the value required to be reset??

File: sources/packages/backend/apps/api/src/route-controllers/student/student.aest.controller.ts [hunk 7]
```diff
@@ -347,6 +352,23 @@ export class StudentAESTController extends BaseController {
...
       throw new NotFoundException("Student does not exists.");
     }
+
+    let isDuplicate = false;
+    if (!payload.confirmDuplicateSIN) {
+      // Check for duplicate SIN if confirmation is not provided.
+      isDuplicate = await this.sinValidationService.checkDuplicateSIN(
+        studentId,
+        payload.sin,
+      );
```
Comment (dheepak-aot) [inline sources/packages/backend/apps/api/src/route-controllers/student/student.aest.controller.ts]: The error can be thrown inside the if block as it does not have reason to live outside. 

```ts
if (!payload.confirmDuplicateSIN) {
      // Check for duplicate SIN if confirmation is not provided.
      const isDuplicate = await this.sinValidationService.checkDuplicateSIN(
        studentId,
        payload.sin,
      );
      if (isDuplicate) {
        throw new UnprocessableEntityException(
          new ApiProcessError(
            "This SIN is currently associated with another student profile and confirmation to allow duplication SIN is missing.",
            SIN_DUPLICATE_NOT_CONFIRMED,
          ),
        );
      }
    }
```

File: sources/packages/web/src/components/common/sin/AddNewSIN.vue (hunk 6 above)
Comment (tiago-graf) [reply sources/packages/web/src/components/common/sin/AddNewSIN.vue]: On cancel we want to reset the entire form to back to original state

**Review by dheepak-aot — APPROVED** [review]
Thanks for making the changes. Looks good 👍
